### PR TITLE
Propagate variable types to generated code to allow statical analysis

### PR DESCRIPTION
**Pull request by MartinMystikJonas**
- new feature
- BC break? no

With this change Latte will propagate known types infomation from ```{varType}```, ```{define}``` and ```{parameters}``` into compiled PHP code in form of ```/** @var type $var */``` annotations. This will allow statical anylysis of resulting code by static analysers like PHPStan, Psalm, etc.

See discussion about this feature here https://github.com/nette/latte/issues/262

### CODE REVIEW

**Comment by MartinMystikJonas** [comment]
@dg Can you please chek this if it is valid approach from your Point of view? It it seems ok to you I will add some more tests for edge cases and support for extracting types from {templateType}

**Comment by MartinMystikJonas** [comment]
@dg Should I consider that like as approval of this approach? If so can you please approve to run workflows? Thanks.

File: src/Latte/Macros/BlockMacros.php [hunk 1]
```diff
@@ -547,7 +552,7 @@ private function addBlock(MacroNode $node, string $layer = null): Block
 	private function extractMethod(MacroNode $node, Block $block, string $params = null): void
 	{
 		if (preg_match('#\$|n:#', $node->content)) {
-			$node->content = '<?php extract(' . ($node->name === 'block' && $node->closest(['embed']) ? 'end($this->varStack)' : '$this->params') . ');'
+			$node->content = '<?php ' . ($node->name === 'block' && $node->closest(['embed']) ? 'extract(end($this->varStack));' : $this->getCompiler()->paramsExtraction)
```
Comment (MartinMystikJonas) [inline src/Latte/Macros/BlockMacros.php:555]: This change is necessary to keep params extraction inside block same as in top-level. Previous version always extracted all params inside block even when {parameters} is used.

File: src/Latte/Compiler/Compiler.php [hunk 2]
```diff
@@ -174,7 +177,8 @@ private function buildClassBody(array $tokens): string
 		$this->output = &$output;
 		$this->inHead = true;
 		$this->macroNode = new RootNode;
-		$this->htmlNode = $this->context = $this->paramsExtraction = null;
+		$this->htmlNode = $this->context = null;
+		$this->paramsExtraction = $this->defaultParamsExtraction;
```
Comment (MartinMystikJonas) [inline src/Latte/Compiler/Compiler.php:181]: Default params extraction needs to be set in the beggining so macros can extend it with type information (prepend @var annotations)

**Comment by dg** [comment]
Hi, please give me some time, I'm busy this week. 

Otherwise, it's best to do PR against the last released version, because I change and forcepush the master a lot.

**Comment by MartinMystikJonas** [comment]
@dg Understood. Take your time and look at it whenever you will have time to check it. Let me know if you have any questions or requests how to finish this PR. 

I will rebase it onto last release.

**Comment by MartinMystikJonas** [comment]
New PR targeted onto v2.10 here #276
